### ISSUE TITLE
vhost/raid0 pending IO never completes

### ISSUE THREAD

**Issue by gnif** [spdk/spdk#981]
When running a QEMU VM with a vhost disk (scsi or blk) some rare accesses may cause an IO operation to hang and never complete.  This can be reliably triggered by running the HDTune Pro `Random seek / size 8MB` test as shown in the below screenshot.

![image](https://user-images.githubusercontent.com/351151/66529464-49866900-eb4f-11e9-82e8-eb722ac3a09c.png)

Even after stopping this and terminating the application the windows disk active time remains at 100% and when attempting to shut down the VM it hangs waiting for vhost stop.

This seems to only occur when a `bdev_raid0` is in use, using a `bdev_malloc` directly doesn't exhibit this behavior and works as expected indicating the fault likely lies in the raid bdev.

A free 15-day trial version of HDTune Pro can be obtained from https://www.hdtune.com

Here is the backtrace of vhost when attempting to shut down the VM
```
Thread 4 (Thread 0x7f3bb969b700 (LWP 57294)):
#0  futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x55aab3677000 <g_dpdk_sem>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:205
#1  do_futex_wait (sem=sem@entry=0x55aab3677000 <g_dpdk_sem>, abstime=0x0) at sem_waitcommon.c:111
#2  0x00007f3bbae75988 in __new_sem_wait_slow (sem=0x55aab3677000 <g_dpdk_sem>, abstime=0x0) at sem_waitcommon.c:181
#3  0x000055aab34ebdae in vhost_session_send_event (pg=0x55aab4c2b5a0, vsession=0x7f3b9c000c40, cb_fn=0x55aab34f193d <vhost_scsi_stop_cb>, timeout_sec=3, 
    errmsg=0x55aab35ea451 "stop session") at vhost.c:971
#4  0x000055aab34f19f5 in vhost_scsi_stop (vsession=0x7f3b9c000c40) at vhost_scsi.c:1469
#5  0x000055aab34ec21a in _stop_session (vsession=0x7f3b9c000c40) at vhost.c:1113
#6  0x000055aab34ec38d in stop_device (vid=0) at vhost.c:1151
#7  0x000055aab34f4723 in spdk_extern_vhost_pre_msg_handler (vid=0, _msg=0x7f3bb969a8b0) at rte_vhost_compat.c:109
#8  0x000055aab358b94f in vhost_user_msg_handler (vid=0, fd=17) at /opt/spdk/dpdk/lib/librte_vhost/vhost_user.c:2003
#9  0x000055aab3581cb6 in vhost_user_read_cb (connfd=17, dat=0x7f3b9c000b20, remove=0x7f3bb969ac24) at /opt/spdk/dpdk/lib/librte_vhost/socket.c:298
#10 0x000055aab357d95a in fdset_event_dispatch (arg=0x55aab3660ca0 <vhost_user+8192>) at /opt/spdk/dpdk/lib/librte_vhost/fd_man.c:286
#11 0x000055aab355d0cd in rte_thread_init (arg=0x55aab5032e00) at /opt/spdk/dpdk/lib/librte_eal/common/eal_common_thread.c:164
#12 0x00007f3bbae6cfa3 in start_thread (arg=<optimized out>) at pthread_create.c:486
#13 0x00007f3bbad9d4cf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
Here is the `msg` from frame 7
```
(gdb) print *msg
$2 = {request = VHOST_USER_SET_MEM_TABLE, flags = 1, size = 72, payload = {u64 = 2, state = {index = 2, num = 0}, addr = {index = 2, flags = 0,
      desc_user_addr = 0, used_user_addr = 2147483648, avail_user_addr = 140556025987072, log_guest_addr = 0}, memory = {nregions = 2, padding = 0,
      regions = {{guest_phys_addr = 0, memory_size = 2147483648, userspace_addr = 140556025987072, flags_padding = 0}, {guest_phys_addr = 4294967296,
          memory_size = 13958643712, userspace_addr = 140558173470720, flags_padding = 2147483648}, {guest_phys_addr = 34762391552, 
          memory_size = 134217728, userspace_addr = 140572398452736, flags_padding = 0}, {guest_phys_addr = 0, memory_size = 0, userspace_addr = 0,
          flags_padding = 0}, {guest_phys_addr = 0, memory_size = 0, userspace_addr = 0, flags_padding = 0}, {guest_phys_addr = 0, memory_size = 0,
          userspace_addr = 0, flags_padding = 0}, {guest_phys_addr = 0, memory_size = 0, userspace_addr = 0, flags_padding = 0}, {guest_phys_addr = 0,
          memory_size = 0, userspace_addr = 0, flags_padding = 0}}}, cfg = {offset = 2, size = 0, flags = 0,
      region = "\000\000\000\000\000\000\000\200\000\000\000\000\000\000\000\300\325\177", '\000' <repeats 14 times>, "\001\000\000\000\000\000\000@\003\000
\000\000\000\000\000@\326\177\000\000\000\000\000\200\000\000\000\000\000\000\000\030\b\000\000\000\000\000\000\b\000\000\000\000\000\000\340\217\331\177",
'\000' <repeats 173 times>}}}
```
`vsession`
```
(gdb) print *vsession
$3 = {vdev = 0x55aab50201c8, vid = 0, id = 0, name = 0x7f3b9c000b50 "vhost.0s0", poll_group = 0x55aab4c2b5a0, initialized = true, started = true,
  needs_restart = false, forced_polling = false, mem = 0x7f3b9c008fa0, task_cnt = 1, max_queues = 3, negotiated_features = 4563402758,
  coalescing_delay_time_base = 0, coalescing_io_rate_threshold = 600, next_stats_check_time = 0, stats_check_interval = 33900000, virtqueue = {{vring = {
        desc = 0x7f3af2141000, avail = 0x7f3af2141800, used = 0x7f3af2141940, log_guest_addr = 18020047168, callfd = 41, kickfd = 34, size = 128},
      last_avail_idx = 0, last_used_idx = 0, tasks = 0x20216b50c600, req_cnt = 0, used_req_cnt = 0, irq_delay_time = 0, next_event_time = 0,      
      vring_idx = 0}, {vring = {desc = 0x7f3af2142000, avail = 0x7f3af2142800, used = 0x7f3af2142940, log_guest_addr = 18020051264, callfd = 40,
        kickfd = 35, size = 128}, last_avail_idx = 0, last_used_idx = 0, tasks = 0x20216b4c2d80, req_cnt = 0, used_req_cnt = 0, irq_delay_time = 0,
      next_event_time = 0, vring_idx = 1}, {vring = {desc = 0x7f3af2143000, avail = 0x7f3af2143800, used = 0x7f3af2143940, log_guest_addr = 18020055360,
        callfd = 33, kickfd = 39, size = 128}, last_avail_idx = 61156, last_used_idx = 61155, tasks = 0x20216b479500, req_cnt = 58864, used_req_cnt = 0,
      irq_delay_time = 0, next_event_time = 0, vring_idx = 2}, {vring = {desc = 0x0, avail = 0x0, used = 0x0, log_guest_addr = 0, callfd = 0, kickfd = 0,
        size = 0}, last_avail_idx = 0, last_used_idx = 0, tasks = 0x0, req_cnt = 0, used_req_cnt = 0, irq_delay_time = 0, next_event_time = 0,
      vring_idx = 4294967295} <repeats 253 times>}, tailq = {tqe_next = 0x0, tqe_prev = 0x55aab50201f8}}
```
`vsession->vdev`
```
(gdb) print *vsession->vdev
$5 = {name = 0x55aab4c28bc0 "vhost.0", path = 0x55aab50157a0 "/var/tmp/vhost.0", cpumask = 0x55aab5020340, registered = true,
  backend = 0x55aab364bf40 <spdk_vhost_scsi_device_backend>, coalescing_delay_us = 0, coalescing_iops_threshold = 60000, vsessions = {
    tqh_first = 0x7f3b9c000c40, tqh_last = 0x7f3b9c008cc0}, vsessions_num = 1, active_session_num = 1, pending_async_op_num = 0, tailq = {tqe_next = 0x0,
    tqe_prev = 0x55aab36501d0 <g_vhost_devices>}} 
```

**Comment by jimharris** [comment]
Is this a raid0 with two NVMe bdevs?  If so does this reproduce with just a single NVMe brevity?

**Comment by gnif** [comment]
I have reproduced this with raid0 and two mallocs also indicating the fault doesn't lie in nvme.

**Comment by jimharris** [comment]
Good triage. Thanks.

**Comment by changpe1** [comment]
The debug message "last_avail_idx = 61156, last_used_idx = 61155" indicates that one IO  didn't return to the vhost target, is this happens with only one malloc block device? if so there should be an issue in bdev layer.

**Comment by gnif** [comment]
The test completes fine with a single malloc device

**Comment by changpe1** [comment]
@gnif Thanks for the information.

**Comment by peluse** [comment]
FYI I've got my qemu windows VM setup w/vhost.  Tomorrow I'll see if I can baseline with a malloc bdev and then move on to repro/debug with RAID.  I doubt I'll need it but @gnif if you could provide your qemu startup command line in advance that would be good (in case it just works for me)

**Comment by peluse** [comment]
@gnif can you also provide you raid config info, total size and strip size? I'm able to get a single disk malloc to pass this test just as you did.  However my 2 disk raid0 created with 2x64MB malloc disks using:
'bdev_raid_create -n raid0 -z 512 -r 0 -b "Malloc0 Malloc1"' doesn't even pass the test. It shows 'failed' but no hangs and I can attempt it dozens of times.
![image](https://user-images.githubusercontent.com/4650360/66671448-d9d8b280-ec10-11e9-9853-42a327ccaabd.png)
Also I am booting to a qcow file and attaching the scsi disk as a data disk only. For this test I doubt that makes a difference but it might.
Shoot me your RAID config line and I'll try that.  Will check a few basic things on my end to wrt why the tests fails pretty quickly.

**Comment by peluse** [comment]
FYI I increased the size of my vol to 1GB and the test passes now and doesn't hang.  I also crated this one with 64K strip size.  Tried raw disk and with NFTS.  Also ran all the other extra tests and they all pass.
![image](https://user-images.githubusercontent.com/4650360/66672276-c6c6e200-ec12-11e9-8acb-14a051690cf3.png)

**Comment by peluse** [comment]
@gnif OK, repro w/strip size 512K.  If you have a minute can you try yours with strip size 64 if it's not too much of a pain? I'l start looking at this though, would just be nice to know our setups are behaving the same and that strip size seems to be the key factor

**Comment by peluse** [comment]
@gnif never mind.  I can now repro with various strip sizes. I don't even think it's RAID at this point but still have more poking around to do.  I can repro with our passthru module on top of malloc but only when I require split_on_optimal_io_boundary which is how the raid module assures that it won't get an IO spanning a strip.  The split code in question is therefore the same stuff that Shuhei just fixed the running out of child IOVs issues.  Have a big beer tasting soon so probably won't have an answer until early next week ;)

**Comment by peluse** [comment]
@gnif I believe I have a *possible* fix - needs more testing and Shuhei to review ;) however I'm out of time and it needs cleanup even before posting for review. I should be able to have it up tomorrow though. It's in the same logic, how we adjust after running out of child_iovs.  In the meantime if you just want to get unblocked you can crank BDEV_IO_NUM_CHILD_IOV up to like 128 or something.  In my test env I only hit a high water mark of I think 67 child_iovs needed. I ran real quick with 256 and it worked fine.

Anyway, path will be up soon.  I did my debug/test with a passthru specifying the same split as raid but did just now test with raid and it works too. I ran it about a dozen times

**Comment by gnif** [comment]
Thank you for your hard work, I would love to be involved more but at the moment I am away overseas and do not have access to my development environment. I have reproduced this with various strip sizes starting at 64k all the way up to 512k. I might be seeing these issues more readily as I have isolated two cores using the `isocpus` kernel arg and use these cores only for the `vhost`, as such the Linux scheduler never takes time away from the processes.

I am sorry I can't do more at the moment.

**Comment by peluse** [comment]
Totally understand, you've been super helpful already in nearly root causing these without any debug - much appreciated!

**Comment by peluse** [comment]
@jimharris @shuhei-matsumoto this got pretty hairy pretty quick. My change worked for the windows case and still passed UT but when I started adding another UT I discovered it was totally off base. I won't confuse things explaining how I thought it worked (can do that in person and confuse less people) but right now I'm not sure how I understand how aligning the final iov can make this work at all.  If the original IO has an iovec makeup such that it needs more than 32 child iovecs due to size/starting address and split boundary, how can we ever transfer all of the data?  The entire IO has to "fit" within one execution of _spdk_bdev_io_split().  The outer most while loop submits child IOs at the end then loops back up to build more with the condition that we haven't run out of child IOs. When we do and there's more data to transfer how can cutting the last IOV short does anything more than transfer less data than the IO wants.

I think all I did in my change was to prevent a hang and I suspect a data integrity test would fail.

Maybe a little white boarding on Mon and Tue morn community call would be good unless I'm totally missing the forest for the trees and someone can correct me in an update here (wouldn't be the first time)

**Comment by shuhei-matsumoto** [comment]
@peluse Thank you for your great effort. Will you share the patch you created? Data integrity means T10 DIF? I could not find what is the cause of the issue from your comment yet :-)

**Comment by peluse** [comment]
@shuhei-matsumoto let me wait on posting the patch until I understand things a little more as it could easily just confuse everything.  And by data integrity I didn't mean DIF, I meant if the did test write/read/compare that it would fail.  The test in question just does reads. Maybe you can comment on my last comment real quick? If all of the split IO's are contained within one bdev_io structure and we don't reuse any child_iov positions and the IO we're trying to split needs, lets say 33 because of its length/offset/alignment requirements how is it possible for us to complete the entire data transfer regardless of alignment?

**Comment by shuhei-matsumoto** [comment]
@peluse 
parent_io->u.bdev.split_remaining_num_blocks knows the progress of submissions. if number of input iovs is 33 and size of child_iovs is 32.

The first call of _spdk_bdev_io_split() will submit the first 32 iovs.
Then in _spdk_bdev_io_split_done(), if we know parent_io->u.bdev.split_outstanding is zero and parent_io->u.bdev.split_remaining_num_blocks is not zero, we will submit the last one input iov.

I think this is the basic IO submission flow.
Does this match what you ask?

**Comment by peluse** [comment]
OK, that makes sense but that's not what I was seeing in the test case I was adding which covered the case where we have to remove >1 IOV and shorten the last one (so for example IOV[31] IOV[30] zero'd out and IOV[29} shorten to block alignment. I could very well have had an issue with my test case though.  I'm unable to look at this more until Monday though, thanks for that last comment - makes much more sense now!

**Comment by shuhei-matsumoto** [comment]
@peluse 

Thanks a lot. 
In a _spdk_bdev_io_spilt call(), if we run out child iovs by the last child IO and the last child IO has data less than block size, the last child IO will have no data due to unwinding. Is this what you are saying? I will do more tomorrow too.

**Comment by peluse** [comment]
LOL I totally missed the very last LOC in _spdk_bdev_io_split_done() . I'll update with more specific details on the windows failure, the patch I was going to propose, and the UT cases that I think are missing - tomorrow though.  AZ cardinals won today so now I need to celebrate and watch my wife's team - the Steelers

**Comment by peluse** [comment]
hmmm, I wonder if we missed rewinding the parent iov offset when we have to shorten the last IOV when running out of child iovs...

**Comment by peluse** [comment]
@shuhei-matsumoto OK I think I see the real problem.  In my first change all I did was rework the rewind logic to make it simpler for debug and the problem went away however my change wasn't complete. Forget that change, what's there now (your last fix) is fine there's just one case (well there may be more I guess) that is causing the hang. We, of course, reset iovcnt each time through the outer while loop. If it turns out that the last time through that loop only has 1 IOV and we have to eliminate it because of alignment then we end up issuing a read/write with 0 iovcnt and somebody doesn't like that (don't know if its windows or in SPDK or vhost or what). The simple fix is to check the iovcnt before bumping bdev_io->u.bdev.split_outstanding and sending a r/w with no iovs.  I'l post in a bit but it will take me a little longer to add a UT to cover this case. There are a few other scenarios that I came up with that need a UT also but those can be separate from this.

**Comment by peluse** [comment]
https://review.gerrithub.io/c/spdk/spdk/+/471313

**Comment by jimharris** [comment]
Thanks @peluse.  FYI - scan-build found a problem with that patch - should @gnif still go ahead and test with rev 1 or wait for an updated version?

**Comment by peluse** [comment]
I just updated the patch, the fix is not 100% complete.  It will prevent the hang but I don't believe will perform the full data transfer in this seemingly very rare condition that happens a lot with this test.  I'll keep the patch updated with status and post a note here when its ready

**Comment by shuhei-matsumoto** [comment]
@peluse 

Thank you for your great effort. I replied a few comments to your patch. I hope they will be any help.

**Comment by gnif** [comment]
Unfortunately, I won't be able to test for at least another week.

**Comment by peluse** [comment]
No worries, thanks for letting us know...

**Comment by changpe1** [comment]
@peluse @shuhei-matsumoto I think Shuhei's commit 3deaf00 has fixed the issue, isn't it?

**Comment by gnif** [comment]
@changpe1 no, this does not fix the issue, that fixed #979 which is a separate issue I reported.

**Comment by changpe1** [comment]
@gnif I see, thanks.

**Comment by peluse** [comment]
fix is ready and has one +2. Needs to be verified by submitter but I think its OK to merge, also has UT with it https://review.gerrithub.io/c/spdk/spdk/+/471313

**Comment by jimharris** [comment]
Hi @gnif - just curious, would you have time to test this patch this week?  I agree with Paul that we can probably merge it now, but would feel better with your stamp of approval!

**Comment by gnif** [comment]
I can test it now, I arrived home safely a few hours ago. I will update this once I have a verdict.

**Comment by gnif** [comment]
@peluse thank you kindly for your hard work, this was clearly a hard one to debug and find an appropriate solution. @jimharris This has solved the issue and I am not seeing any negative effects from it.

**Comment by jimharris** [comment]
Patch merged.  Thanks @peluse!
